nexus/transport.go

```diff
@@ -0,0 +1,707 @@
+package nexus
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"maps"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+)
+
+const getResultContextPadding = time.Second * 5
+
+var errOperationWaitTimeout = errors.New("operation wait timeout")
+
+type (
+	// Transport is the low-level abstraction used by [Client] and [OperationHandle] to make network calls.
+	//
+	// NOTE: Experimental
+	Transport interface {
+
+		// StartOperation calls the configured Nexus endpoint to start an operation.
+		//
+		// This method has the following possible outcomes:
+		//
+		//  1. The operation completes successfully. The result of this call will be set as a [LazyValue] in
+		//     StartOperationResponse.Complete and must be consumed to free up the underlying connection.
+		//
+		//	2. The operation completes unsuccessfully. The response will contain an [OperationError] in
+		//	   StartOperationResponse.Complete
+		//
+		//  3. The operation was started and the handler has indicated that it will complete asynchronously. An
+		//     [OperationHandle] will be returned as StartOperationResponse.Pending, which can be used to perform actions
+		//     such as getting its result.
+		//
+		//  4. There was an error making the call. The returned response will be nil and the error will be a
+		//     [HandlerError].
+		//
+		//  5. Any other error.
+		//
+		// NOTE: Experimental
+		StartOperation(
+			ctx context.Context,
+			service string,
+			operation string,
+			input any,
+			options StartOperationOptions,
+		) (*StartOperationResponse[*LazyValue], error)
+
+		// GetOperationInfo returns information about a specific operation. Should only be called through an
+		// [OperationHandle].
+		//
+		// NOTE: Experimental
+		GetOperationInfo(
+			ctx context.Context,
+			service string,
+			operation string,
+			token string,
```

Comment:
I think we may hit a situation where transport call options are not necessarily the same as their high-level counterparts. These 3 are transport-only options. But they are not in a structure where we can ever add more transport-only options, so they are the last 3 ever. I think we should consider structures specific to the transport calls (can prefix their to delineate) even at the acceptable cost of more types/conversion. The transport options can carry the high-level options struct as a field, don't have to expand out field for field.